feat: igraph_product() computes the graph product of two graphs

gulshan-123 opened the pull request:
<!-- Please disclose any use of LLMs (ChatGPT, Copilot, Gemini, etc.) during the preparation of this PR. -->

<!-- The following text must be left intact and the box checked before the PR can be merged. -->

- [x] By submitting this pull request, I assign the copyright of my contribution to _The igraph development team_.

Partially Fixes #2729 

To Implement:
- [x] [Cartesian Product](https://en.wikipedia.org/wiki/Cartesian_product_of_graphs)
- [x] [Tensor Product](https://en.wikipedia.org/wiki/Tensor_product_of_graphs)

gulshan-123 commented:
Hello @szhorvat ,

I have defined an enum for the `igraph_product_t` type. The `igraph_product` function will internally call the appropriate function such as `igraph_i_cartesian` based on the selected enumeration value.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,47 @@
+#include "math/safe_intop.h"
+
+
+igraph_error_t igraph_i_cartesian_product(igraph_t *res,
```
Please make this `static` and drop `igraph_i_` from the name.

We use:

 - `igraph_` for public functions
 - `igraph_i_` for non-public functions that are accessible from multiple compilation units
 - No prefix for `static` functions accessible only from the current compilation unit

szhorvat commented on include/igraph_operators.h:
```diff
@@ -40,6 +40,10 @@ __BEGIN_DECLS
 /* Graph operators                                    */
 /* -------------------------------------------------- */
 
+typedef enum {
+        IGRAPH_CARTESIAN_PRODUCT,
+} igraph_product_t;    
+
```
Please put this in `igraph_constants.h`, add `= 0` for predictability, and name all values as `IGRAPH_PRODUCT_XXX`, i.e. `IGRAPH_PRODUCT_CARTESIAN`.

gulshan-123 commented:
There are two options if one of graph is directed but the other is not.
- We may raise an Error
- We may assume each undirected edge u--v as u->v plus v->u
I think the first option is good, as done by NetworkX: [cartesian-product](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.operators.product.cartesian_product.html)
I am unable to think scenario where second will be beneficial

szhorvat commented:
> * We may raise an Error

I vote for this.

gulshan-123 commented:
Hi,

In our implementation of the Cartesian product for multigraphs, we're handling self-loops as follows:

 If the first graph has k self-loops and the second graph has l self-loops (both being 1-vertex multigraphs), then the Cartesian product results in 1 vertex and k + l self-loops.

This is because each self-loop in the first graph contributes a self-loop in the product (with the vertex from the second graph fixed), and vice versa. So, both sets of self-loops are preserved and added independently.

In contrast, I noticed that NetworkX behaves differently. For the same inputs:

It gives only 1 self-loop when both graphs have 1 self-loop.
In general, it returns max(k, l) self-loops, not k + l.

szhorvat commented:
I will only be able to look at this later, but general principles we use to make a decision on such things:

 - Do the usual relations still hold for multigraphs/loopy graphs? For example, see the formulas for the edge count here: https://en.wikipedia.org/wiki/Graph_product
 - Is there an equivalence between using undirected graphs, or the directed counterpart where each undirected edge is replaced by two reciprocal directed ones?

There is often no single correct answer, but a compromise needs to be made and documented.

Again, I have not looked at this at all yet, I'm just giving suggestions on what to consider.

gulshan-123 commented:
Our implementation is correct per the edge counts on Wikipedia. I'll add docs and then proceed with the tensor product.

gulshan-123 commented:
Hi @szhorvat,
I have added support for the Cartesian product and have started working on the tensor product implementation. (Should I create separate PR for tensor product?) Whenever you have some time, I would greatly appreciate it if you could review the changes.

Thank you!

szhorvat commented:
I probably won't have time to review this today. So that I don't hold you up, feel free to go ahead with the tensor product in the same PR.

szhorvat commented:
Due to many outstanding PRs, I'll look at this later.

Is the tensor product considered complete now?

Can you please put all graph product tests into a single file? There is some cost to having a large number of test files (notably, slow test build times). 

I'll add a few minor comments, but these will be localized nitpicks not a proper review. I'll also request a review from Copilot which recently gained C support. Keep in mind that such AI tools often give incorrect suggestions! I am just curious how well it works. If it does find issues which in your judgement are real problems, do fix them. But don't fix things that are not necessarily real problems just to satisfy the AI.

szhorvat commented on include/igraph_constants.h:
```diff
@@ -220,6 +220,10 @@ typedef enum { IGRAPH_ROW_MAJOR = 0,
                IGRAPH_COLUMN_MAJOR = 1
              } igraph_matrix_storage_t;
 
+typedef enum { IGRAPH_PRODUCT_CARTESIAN = 0,
+               IGRAPH_PRODUCT_TENSOR = 1,
+             } igraph_product_t;
```
```suggestion
typedef enum { IGRAPH_PRODUCT_CARTESIAN = 0,
               IGRAPH_PRODUCT_TENSOR,
             } igraph_product_t;
```

Stylistic nitpick.

szhorvat commented on interfaces/functions.yaml:
```diff
@@ -2080,6 +2080,11 @@ igraph_induced_subgraph_map:
         OPTIONAL OUT INDEX_VECTOR map, OPTIONAL OUT INDEX_VECTOR invmap
     DEPS: vids ON graph
 
+# default product type=Cartesian
```
```suggestion
```

Let's remove the redundant comment.

Using `CARTESIAN` below is good. This value will be associated with the enum value on the C side in each high-level interface that makes use of the interface definition file.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Cartesian product of directed and undirected graphs is not supported. "
+                   "You can convert an undirected graph to a directed one using igraph_to_directed, "
+                   "or a directed graph to an undirected one using igraph_to_undirected.", IGRAPH_EINVAL);
```
Instead of saying that this is not supported, let's say that the operation is invalid.

Let's not refer to C-specific function names in error messages that will be shown in various non-C contexts (within Python, R, Mathematica).

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   IGRAPH_SAFE_MULT(vg1, eg2, &eres);
+   IGRAPH_SAFE_MULT(vg2, eg1, &temp);
+   IGRAPH_SAFE_ADD(eres, temp, &eres);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 2*eres);
```
`2*eres` may still overflow.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   igraph_integer_t edge_index = 0;
+   // Edges from g1
+   for (i = 0; i < eg1; ++i) {
+      igraph_edge(g1, i, &from, &to);
```
Avoid this function and use the `IGRAPH_FROM` and `IGRAPH_TO` macros instead. They have better performance as they avoid function calls.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+
+   // Edges from g2
+   for (i = 0; i < eg2; ++i) {
+      igraph_edge(g2, i, &from, &to);
```
Use IGRAPH_TO/FROM.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   igraph_integer_t vg2 = igraph_vcount(g2);
+   igraph_integer_t eg1 = igraph_ecount(g1);
+   igraph_integer_t eg2 = igraph_ecount(g2);
+   igraph_integer_t vres;
+   igraph_integer_t eres;
```
Let's use more descriptive names such as `vcount` and `ecount`.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Tensor product of directed and undirected graphs is not supported. "
+                   "You can convert an undirected graph to a directed one using igraph_to_directed, "
+                   "or a directed graph to an undirected one using igraph_to_undirected.", IGRAPH_EINVAL);
```
Same comment as in the Cartesian product error message.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,217 @@
+   igraph_integer_t edge_index = 0;
+
+   for (i = 0; i < eg1; ++i) {
+      igraph_edge(g1, i, &from1, &to1);
+
+      for (j = 0; j < eg2; ++j) {
+         igraph_edge(g2, j, &from2, &to2);
```
As before, use IGRAPH_TO/FROM.

szhorvat commented:
Be sure to add a newline at the end of the file so all CI tests pass.

gulshan-123 commented:
> Is the tensor product considered complete now?

Yes, the implementation is complete. I will now add tests for directed graphs and update the documentation.

> Can you please put all graph product tests into a single file? There is some cost to having a large number of test files (notably, slow test build times).

Thank you for the review! I will consolidate the tests into a single file and push the changes shortly.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_operators.h"
+
+#include "igraph_conversion.h"
+#include "igraph_interface.h"
+
+#include "core/interruption.h"
+#include "math/safe_intop.h"
```
The convention we use for header includes is:

 1. First comes the header where the functions defined in this file are declared.
 2. This is followed by a newline, then public headers alphabeticaly.
 3. A newline, then private headers alphabetically.
 4. Finally, any standard library headers comes last, though these are rarely needed.

Can you please reorder these and remove unused headers? A good IDE helps you determine what is unused, although sometimes they give incorrect information. I believe `#include "igraph_conversion.h"` is not needed here.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Cartesian product between a directed and an undirected graph is invalid. "
+         "Please ensure that both graphs have the same directionality.", IGRAPH_EINVAL);
+   }
```
```suggestion
   if (directed1 != directed2) {
      IGRAPH_ERROR("Cartesian product between a directed and an undirected graph is invalid.",
                   IGRAPH_EINVAL);
   }
```

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+   igraph_integer_t from, to;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
```
This should be included in the documentation.

Let's document the vertex ordering for all product types, and keep them consistent with each other.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+ * directed graph is required, the undirected graph can be converted to a directed graph using \ref 
+ * igraph_to_directed, or the directed graph can be converted to an undirected graph using \ref 
+ * igraph_to_undirected.
+ * </para><para>
```
```suggestion
 *
```

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+/**
+ * \function igraph_product
+ * \brief Computes the graph product of two graphs, based on the specified product type.
+ * </para><para>
```
Be sure to have an empty line before paragraphs so that the _code_ version is readable.

Please add `\experimental` at the beginning of the docs. By policy, all new functions start out as experimental.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+ * </para><para>
+ * Supported graph product types are:
+ * \clist
+ *    \cli IGRAPH_PRODUCT_CARTESIAN
```
Please also include docs for the tensor product.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+ * \endclist
+ * Both graphs must be of the same type, either directed or undirected. If a product of an undirected and a
+ * directed graph is required, the undirected graph can be converted to a directed graph using \ref 
+ * igraph_to_directed, or the directed graph can be converted to an undirected graph using \ref 
```
Conventionally, when referring to functions, we always add parentheses.

Use `\ref fun()` instead of `\ref fun`.

szhorvat commented on src/operators/products.c:
```diff
@@ -0,0 +1,221 @@
+   //   will have new vertex id: i * vcountg2 + j
+   igraph_integer_t edge_index = 0;
+
+   for (i = 0; i < ecountg1; ++i) {
```
Let's aim to declare iterator variables within `for (...` directly. Making variable scope as local as possible helps readability.

It is only in rare cases that this is not possible or not beneficial to readability.

szhorvat commented on tests/unit/igraph_product.c:
```diff
@@ -0,0 +1,243 @@
+    igraph_vector_bool_init(&periodic, 2);
+    VECTOR(dims)[0] = 4;
+    VECTOR(dims)[1] = 3;
+    VECTOR(periodic)[0] = 1;  // periodic in first dim (C4)
+    VECTOR(periodic)[1] = 0;  // non-periodic in second dim (P3)
```
Use `true`/`false`/`NULL` instead of 1/0/1 for all booleans/pointers.

szhorvat reviewed (changes requested):
Please use 4 spaces for indentation instead of 3.

Some of the comments are nitpicks, but sticking to the preferred style will be helpful in the long term and make it easier to work together.

gulshan-123 commented:
@szhorvat , Thank you for the time. Somehow VS Code was using 3 indents in that file. I have corrected it now.

I have now added the following:
1. Docs for tensor product
2. index calculation in new graph as i*V2+j, added in docs
4. Headers reordering
5. consistent style

Please tell if more changes are required.

szhorvat reviewed (approved):
Thank you!

gulshan-123 commented:
> Thank you @gulshan-123 ! How do you want to be credited? We usually use both the full name and the GitHub handle.

Thank you very much! I am honored to have contributed to igraph. Kindly credit me as Gulshan Kumar (@gulshan-123).
Looking forward to contributing further!

ntamas commented:
Looks good to me. Please make sure to update `.all-contributorsrc` (I just use `all-contributors add` from the command line).
